src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,181 @@ QPainterPath QPainter::clipPath() const
     }
 }
 
+/*!
+    Returns the current clip path in logical coordinates, unrounded.
+
+    The main purpose for this function is inside QSvgGenerator,
+    where we do not want any points to be rounded during the
+    creation of the file.
+
+    \warning QPainter does not store the combined clip explicitly as
+    this is handled by the underlying QPaintEngine, so the path is
+    recreated on demandand transformed to the current logical
+    coordinate system.This is potentially an expensive operation.
+
+    \sa setClipPath(), clipRegion(), setClipping()
+*/
```

Comment:
I really think it would be better to put this change in the svg engine rather than here. As it says in the docstring, "QPainter does not store the combined clip explicitly as this is handled by the underlying QPaintEngine". Seeing as you already have changes to that engine, and the bug being fixed is only in that one specific engine, why not just put this there?

Notice how `d->updateState(d->state);` is called at the end of all the `setClipX` methods, with the newly added clip geometry and operation. So you should be able to take that information and update a clipping state you store in the engine, without having to do this big complicated method each time you need to ask for the clipping.

That would also solve the problem of having to set the clipping every time updateState is called when writing the output. Instead, you know the current clipping (or lack of it), so you write your current clip state to the <g> tag if it exists, and only create a new clipping <defs> entry when it actually changes. Similar to how [saveLinearGradientBrush](https://code.woboq.org/qt5/qtsvg/src/svg/qsvggenerator.cpp.html#_ZN15QSvgPaintEngine23saveLinearGradientBrushEPK9QGradient) calls [generateGradientName](https://code.woboq.org/qt5/qtsvg/src/svg/qsvggenerator.cpp.html#_ZN22QSvgPaintEnginePrivate20generateGradientNameEv) to increment the current gradient state and increment a counter if required, before streaming it to `defs` at that point for later output.

Reply:
Even though this isn't called anywhere else so the scope of the change is limited by that, seems best to avoid changing this class at all if you don't have to

Reply:
Doesn't this method have to be here because of private d-> references and those can only be accessed from the QPainterClass class?

Reply:
Nah, it's only `d->state`, which is passed to the engine's `updateState` method anyway